add `--port` parameter for rasa interactive

pnandhini11 opened the pull request:
**Proposed changes**:
- Added --port parameter for the user to give any port for Rasa interactive learning.  This parameter will be used to start core and followed by interactive learning. 
- close https://github.com/RasaHQ/rasa/issues/4709

**Status (please check what you already did)**:
- [x ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4866) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4866) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=4866) it.</sub>

erohmensing commented:
Thanks for the PR! We'll give it a review as soon as possible.

pnandhini11 commented:
> Thanks for the PR! We'll give it a review as soon as possible.

Hi @erohmensing ,  This is the first time I am trying to contribute for an issue.  Not sure if I have followed all the standards correctly.

erohmensing commented:
@pnandhini11 so far things look pretty good! A few changes I can see here:
*  reformat files using black (please check (Readme)[https://github.com/RasaHQ/rasa_nlu#code-style] for instructions)
*  update the changelog

pnandhini11 commented:
> @pnandhini11 so far things look pretty good! A few changes I can see are still necessary:
> 
> * reformat files using black (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)
> * update the changelog
> 
> For this PR, I don't think an automated test or updates to the documentation is necessary.

Do I have reformat and checkin again?

erohmensing commented:
What do you mean? For reformatting, you should run black via the instructions, and then push any changes that are made to the files.

pnandhini11 commented:
@erohmensing  Hi, I am using windows system and not able to run pytype from windows. Can you help me solve this check?  Also you mentioned automated test not required but build failed in test as well.

erohmensing commented:
I think this is a good place for your reviewer @ArjaanBuijk to step in to help you get the tests passing 🙂

ArjaanBuijk commented:
Hi @erohmensing , thanks for the reviews so far.
Hi @pnandhini11 , I will have a look at the tests to check why they fail.

ArjaanBuijk commented:
Hi @pnandhini11 ,

A few things need to be fixed before your update can be merged. The tests fail, and you can view what tests by clicking on the `Details` link next to the **Tavis CI - Pull Request**:
![image](https://user-images.githubusercontent.com/16159200/70160802-15c84b00-1689-11ea-9cef-9b7bb646c7ba.png)

This brings you to the page where it lists the tests that fail, and you can click to inspect the details:
![image](https://user-images.githubusercontent.com/16159200/70160953-532cd880-1689-11ea-8ea1-9b1c6ebea559.png)

The best way to make these changes and then verify that all is OK, is to run the tests locally. It is described in the README file how to do this. See these sections:

- [Type Checking](https://github.com/RasaHQ/rasa#type-checking)
- [Running the tests](https://github.com/RasaHQ/rasa#running-the-tests)

I had a look at the failing tests and these functions need to be updated:

- rasa/cli/arguments/interactive.py: 
    - **line 44**, in set_interactive_arguments: Function add_port_param was called with the wrong arguments [wrong-arg-types]
  Expected: (parser: argparse.ArgumentParser)
  Actually passed: (parser: argparse._ArgumentGroup)
    - **line 67**, in set_interactive_core_arguments: Function add_port_param was called with the wrong arguments [wrong-arg-types]
  Expected: (parser: argparse.ArgumentParser)
  Actually passed: (parser: argparse._ArgumentGroup)

- tests/cli/test_rasa_interactive.py::test_interactive_core_help
- tests/cli/test_rasa_interactive.py::test_interactive_help

Let me know if it is clear what to do.

pnandhini11 commented:
> @pnandhini11 , pease see my comments about the failing tests.

Hi @ArjaanBuijk , @erohmensing ,
    Thanks for the reply.  I couldn't use Pycheck as I am having windows system and no pycheck for windows yet.  Thats why I needed your help. 

Have a question on this  : 

(Copy from your previous message ) 

rasa/cli/arguments/interactive.py:

**line 44, in set_interactive_arguments: Function add_port_param was called with the wrong arguments [wrong-arg-types]
Expected: (parser: argparse.ArgumentParser)
Actually passed: (parser: argparse._ArgumentGroup)
line 67, in set_interactive_core_arguments: Function add_port_param was called with the wrong arguments [wrong-arg-types]
Expected: (parser: argparse.ArgumentParser)
Actually passed: (parser: argparse._ArgumentGroup)**


My Question :  I am not sure why to use _ArgumentGroup to pass the parameters to add_port_param method.  I have made it using (parser: argparse.ArgumentParser) because to be inline with existing code followed in the entire repository. 


You can check the existing argument coded in interactive.py : 

def add_skip_visualization_param(parser: argparse.ArgumentParser):
    parser.add_argument(
        "--skip-visualization",
        default=False,
        action="store_true",
        help="Disable plotting the visualization during interactive learning.",
    )

Let me know how to go about it now. Thanks in advance.

pnandhini11 commented:
@erohmensing @ArjaanBuijk  any updates to my previous comment?

ArjaanBuijk commented:
@pnandhini11 , sorry for delay in response. I will be able to review the issues later today and will then get back to you on them.

ArjaanBuijk commented:
@wochinge ,

Is there a reason that `rasa/cli/arguments/interactive.py` is so different from `rasa/cli/arguments/shell.py` ?

I see two ways to implement the `--port` parameter for `rasa interactive`:

1. Just add the `--port` parameter as @pnandhini11 has implemented.
2. from `rasa/cli/arguments/interactive.py` call `add_server_arguments` , as is done in `rasa/cli/arguments/shell.py`

wochinge commented:
> Is there a reason that rasa/cli/arguments/interactive.py is so different from rasa/cli/arguments/shell.py ?

`shell` is just speaking to a trained model. `interactive` is speaking to trained model, but also training a model and also visualizing things.

> Just add the --port parameter as @pnandhini11 has implemented.
from rasa/cli/arguments/interactive.py call add_server_arguments , as is done in rasa/cli/arguments/shell.py

How about a middle course and 
- extract the current port argument declaration from `add_server_arguments` to a `add_port` function
- calling `add_port` in `add_server_arguments` and `set_interactive_arguments`

wochinge commented:
Maybe we can also de-duplicate `set_interactive_arguments` and `set_interactive_core_arguments` a bit

ArjaanBuijk commented:
Hi @pnandhini11 , would you be able to implement the functionality using an `add_port` function as @wochinge proposes? I like that approach a lot.

pnandhini11 commented:
> Hi @pnandhini11 , would you be able to implement the functionality using an `add_port` function as @wochinge proposes? I like that approach a lot.

Let me work on this and get back to you.

wochinge commented:
@pnandhini11 What's the state? Can we somehow help you?

pnandhini11 commented:
Hi @wochinge , I made the updates as discussed but got busy with my work. And missed updating the status. Will do it by this weekend.

wochinge commented:
@pnandhini11 No rush, I was just curious about the status 👍

pnandhini11 commented:
@ArjaanBuijk @wochinge Made changes in rasa\cli\arguments\interactive.py to call add_server_argument instead of add_port method i created last. 
Check and let me know if i understood your idea correctly. 
By mistake i deleted rasa\cli\interactive.py and added again in this commit. Didn't do any code changes in it.

wochinge commented on rasa/cli/arguments/interactive.py:
```diff
@@ -40,6 +42,7 @@ def set_interactive_arguments(parser: argparse.ArgumentParser):
     add_dump_stories_param(train_arguments)
     add_force_param(train_arguments)
     add_persist_nlu_data_param(train_arguments)
+    add_server_arguments(parser)
```
I think we should not add the other server arguments, and then not using them. I suggest just adding the `--port` argument for now

pnandhini11 replied:
ok.

wochinge commented on rasa/core/training/interactive.py:
```diff
@@ -1444,9 +1444,13 @@ def _print_help(skip_visualization: bool) -> None:
 
+    if server_args.get("port"):
+        port = server_args.get("port")
+        DEFAULT_SERVER_URL = "{}://localhost:{}".format("http", port)
```
if http is hardcoded we can just do
```suggestion
        DEFAULT_SERVER_URL = f"http://localhost:{port}"
```

wochinge commented on rasa/core/training/interactive.py:
```diff
@@ -1467,7 +1471,10 @@ def _serve_application(app, stories, skip_visualization):
     update_sanic_log_level()
 
-    app.run(host="0.0.0.0", port=DEFAULT_SERVER_PORT)
+    if port:
```
a bit shorter:

```python
port = port or DEFAULT_SERVER_PORT
app.run(host="0.0.0.0", port=port)
```

wochinge commented on rasa/core/training/interactive.py:
```diff
@@ -1444,9 +1444,13 @@ def _print_help(skip_visualization: bool) -> None:
+def _serve_application(app, stories, skip_visualization, server_args):
     """Start a core server and attach the interactive learning IO."""
 
+    if server_args.get("port"):
```
otherwise the variable might not be defined when running the server
```suggestion
    port = None
    if server_args.get("port"):
```

wochinge commented on rasa/core/training/interactive.py:
```diff
@@ -1589,7 +1599,7 @@ def run_interactive_learning(
         )
 
-    _serve_application(app, stories, skip_visualization)
+    _serve_application(app, stories, skip_visualization, server_args)
```
let's pass in the `port` instead of `server_args` so we don't have to repeat the unpacking.

pnandhini11 replied:
ok. Though it will be useful in future to any new argument. Anyways will pass port directly and do the necessary changes.

wochinge replied:
> Though it will be useful in future to any new argument.

What do you mean?

wochinge commented:
@pnandhini11 How is it going?

tmbo commented:
wrapped up in https://github.com/RasaHQ/rasa/pull/5837
